pkg/process/util/api/weighted_queue.go

```diff
@@ -69,41 +76,29 @@ func (q *WeightedQueue) Weight() int64 {
 	return q.currentWeight
 }
 
-// Poll retrieves the head of the queue or blocks until an item is available.  The provided exit channel can be closed
-// to interrupt the blocking operation.  Returns the head of the queue and true or nil, false if the poll was
-// interrupted by the closing of the exit channel
-func (q *WeightedQueue) Poll(exit chan struct{}) (WeightedItem, bool) {
-	for {
-		select {
-		case <-exit:
-			return nil, false
-		default:
-
-		}
-
-		q.mu.Lock()
+// Poll retrieves the head of the queue or blocks until an item is available or
+// the WeightedQueue is stopped.  Returns the head of the queue and true or
+// nil, false if the WeightedQueue is stopped.
+func (q *WeightedQueue) Poll() (WeightedItem, bool) {
+	q.mu.Lock()
 
-		// If the queue is empty, wait for a signal on the dataAvailable channel
-		if q.queue.Len() == 0 {
-			q.mu.Unlock()
+	// If the queue is empty and we aren't stopped, wait for a signal
+	for q.queue.Len() == 0 && !q.stop {
+		q.cv.Wait()
+	}
 
-			select {
-			case <-q.dataAvailable:
-				continue
-			case <-exit:
-				return nil, false
-			}
-		}
+	if q.stop {
+		return nil, false
```

Comment:
Wouldn't this return leave the lock locked? Wouldn't another routine/thread coming into this block indefinitely due to that?

Reply:
Oh, indeed!